src/functions/assert/Time/Assert-Before.ps1

```diff
@@ -1,41 +1,51 @@
 ﻿function Assert-Before {
     [System.Diagnostics.CodeAnalysis.SuppressMessageAttribute('PSUseProcessBlockForPipelineCommand', '')]
+    [CmdletBinding(DefaultParameterSetName = "Now")]
     param (
-        [Parameter(Position = 1, ValueFromPipeline = $true)]
+        [Parameter(Position = 2, ValueFromPipeline = $true)]
         $Actual,
-        [Parameter(Position = 0, ParameterSetName = "Expected")]
-        [DateTime] $Expected,
 
-        [Parameter(Position = 0, ParameterSetName = "Ago")]
-        $TimeAgo,
+        [Parameter(Position = 0, ParameterSetName = "Now")]
+        [switch] $Now,
+
+        [Parameter(Position = 0, ParameterSetName = "Fluent")]
+        $Time,
+
+        [Parameter(Position = 1, ParameterSetName = "Fluent")]
+        [switch] $Ago,
+
+        [Parameter(Position = 1, ParameterSetName = "Fluent")]
+        [switch] $FromNow,
```

Comment:
I was thinking about that, but that gives some generic "cannot figure out parameter set" error, no?

Reply:
If you call `Assert-After -Time 1s -Ago -FromNow` yes.

However tab completion will hide one when the other is already used and get-help will clearly show them in different sets.

Reply:
Maybe just skip the `-Ago/FromNow` and add support for negative value in string pattern? It simplifies both the syntax. E.g.

```powershell
Assert-After [-Now] [-Actual] <Object> [<CommonParameters>]
Assert-After [-TimeSpanFromNow] <timespan> [-Actual] <Object> [<CommonParameters>]
Assert-After [-TimePatternFromNow] <string> [-Actual] <Object> [<CommonParameters>]
Assert-After [-Expected] <datetime> [-Actual] <Object> [<CommonParameters>]
```

I'm a little conflicted with the string syntax in general. I really appreciate the readability and efficiency for very simple timespans and I'm fine with keeping it. However, it is proprietary and currently doesn't support for negative values and multiple units (see below).

TimeSpan itself has a string pattern that is applicable for timespans everywhere in PowerShell. PowerShell is all about learning things once, so we should promote the native ways when possible.
- '1:15:10' -> 1h15m10s # currently not supported in Fluent, would be 4510s
- -1.0:0:0' -> -1d # Fluent syntax is nicer, but `New-TimeSpan -Day -1` is also a readable alternative

Reply:
I was not sure if I am going too deep with the syntax. I agree on switching to different parameter sets. Your point about negative, multiple units and learning syntax once is all valid. But does it need to be that complicated? 

What I was envisioning here is that the text syntax is used for more general use cases, like: 
- I fetched all data, and user provided 1 week filter, so no item should be older than a week. 
- I just created a user, make sure it is not older than 30 seconds. 
- I run my code, make sure it completes within 1 second. 

Usually when I need a super precise time, I am just converting it from some other event that already has some timespan or a date, instead of relating it to now, so that syntax would not be usable there.

But I will think about it. I probably also need to capture the `$currentTime` in begin { } block, so it starts before the command. So e.g. if there is a user that is exactly 1 week old we don't fail.

Reply:
note: 
```powershell
function a {
    param ([Parameter(valuefrompipeline=$true)][object[]]$a)
    begin {
        $currentTime = [datetime]::UtcNow.ToLocalTime()
    }

    end {
        [datetime]::UtcNow.ToLocalTime() - $currentTime
    }
}


Get-Command | a
```

Reply:
> What I was envisioning here is that the text syntax is used for more general use cases

I agree with your vision here. I do believe the parameter sets I suggested + supporting negative values in the Fluent-pattern would be a balanced solution. 🙂 

- Short and simple for most use cases with Fluent
- Advanced with datetime or timespan. _Timespan is not really required, but should stay as long as Fluent is there_
- No switches as they clutter what a `-` could provide
- No ambiguous object-parameters.

Reply:
It’s also seems that it would be useful to provide a date time and timespan in the string format to be able to provide a $now. E.g. when you run a command and need to ensure the items you grabbed are not a week old since before the command started, and not since after the command finished.

Reply:
Not sure I follow. `Assert-After -Ago "1w"`?

Reply:
$now = datetime.now
$files = Get-ReportsButNotOlderThan1Week
… | Should-BeAfter -Time $now -Ago 1w

so if the get-reports… command takes 30 seconds to finish and the oldest report is exactly one week old the assertion wont fail

Reply:
Ah okay. That's the "advanced" case I expected `Should-BeAfter -Expected $now.AddDays(-7)` to solve.

But I see your point, the datetime reference could easily be a parameter for timespan and string sets  :)